### PR TITLE
Adds type traits and overload for multiplication of eigen complex types

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This adds `multiply` overloads so eigen matrices with complex types can use `multiply`

## Tests

Tests were added to test evaluation of combinations of complex and non-complex matrix types.

```
./runTests.py ./test/unit/math/mix/fun/multiply_test
```

## Side Effects

We still can't do `operator*` for complex matrix types (or vars I think) I have another PR that allows for eigen var matrices but that would be a seperate PR

## Release notes

Add support for multiplication of complex eigen matrices with inner autodiff types

## Checklist

- [x] Math issue #2003

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
Also I think some other errors will pop up in the future since right now in `rev` we have a lot of requires like `require_any_var_t<T1, T2>`, but those specializations are really for vars and doubles. We will probably need to do something to make sure `prim` overloads are used when any type is complex.

**Comment by SteveBronder** [comment]
@t4c1 hang off on merging this. I'm testing it out upstream to make sure it has all the stuff we need. We may need to have const and non-const versions of some of these (at least I'm getting "as 'this' argument discards qualifiers in some places)

File: stan/math/rev/fun/multiply.hpp
Comment (bbbales2) [inline stan/math/rev/fun/multiply.hpp:557]: Was it a problem if this was taking higher order autodiff types or something?

File: stan/math/rev/fun/multiply.hpp
Comment (bbbales2) [inline stan/math/rev/fun/multiply.hpp:596]: Why is this require_any_eigen_vt required?

**Review by bbbales2 — CHANGES_REQUESTED** [review]
Review!

File: test/unit/math/mix/fun/multiply_test.cpp
Comment (SteveBronder) [reply test/unit/math/mix/fun/multiply_test.cpp:98]: Yeah so previously these just did `auto d_d_mat = Matrix<> * Matrix<>` but the auto there is just a `Product<Matrix<>, Matrix<>>` that's not evaluated. The complex PR I have up fixes this all the way through tho'

File: stan/math/prim/fun/value_of.hpp
Comment (SteveBronder) [reply stan/math/prim/fun/value_of.hpp:28]: Ah, 2 things

(1) this needs to be made into just `auto` and not `decltype(auto)` (see [this](https://github.com/stan-dev/math/pull/1844/files/e355481962c8b2537062be00a1dc2078a52d0918#r486852747) convo)
(2) `_st_` checks on the `scalar_type` and for complex types that's defined as 

```cpp
template <typename T>
struct scalar_type<T, std::enable_if_t<is_complex<T>::value>> {
  using type = std::complex<typename std::decay_t<T>::value_type>;
};
```

File: stan/math/rev/fun/Eigen_NumTraits.hpp
Comment (SteveBronder) [reply stan/math/rev/fun/Eigen_NumTraits.hpp:37]: Yeah I think this should be double

File: stan/math/prim/fun/multiply.hpp
Comment (SteveBronder) [reply stan/math/prim/fun/multiply.hpp:94]: To use regular product here we would need to specialize `general_matrix_matrix_product` and `general_matrix_vector_product` in `Eigen_NumTraits.hpp`. `lazyProduct()` is inefficient but handles promotion nicer so that complex matrices and regular matrices work. tbh I just didn't want to write the gemm specializations because they are complicated but in the future we can do that

File: stan/math/rev/fun/multiply.hpp
Comment (SteveBronder) [reply stan/math/rev/fun/multiply.hpp:557]: This was just a cleanup because the multiply requires are very specialized and confusing

File: stan/math/rev/fun/multiply.hpp
Comment (SteveBronder) [reply stan/math/rev/fun/multiply.hpp:596]: Let me rethink this. The logic here is that we want to accept combinations of vars and arithmetic types but at least one of the types has to have an var scalar.

File: stan/math/prim/fun/multiply.hpp
Comment (bbbales2) [inline stan/math/prim/fun/multiply.hpp:96]: [Optional] Get rid of .eval()

File: stan/math/prim/fun/multiply.hpp
Comment (bbbales2) [inline stan/math/prim/fun/multiply.hpp:153]: I like `require_all_not_st_var` honestly

File: test/unit/math/mix/fun/multiply_test.cpp
Comment (bbbales2) [inline test/unit/math/mix/fun/multiply_test.cpp:238]: In terms of combinations this is missing rowvec/mat, mat/rowvec, but I am convinced it works.

**Review by bbbales2 — COMMENTED** [review]
Optional comments, I'll approve and merge when tests pass

File: stan/math/prim/fun/multiply.hpp
Comment (SteveBronder) [reply stan/math/prim/fun/multiply.hpp]: I switched this to `require_not_var_t<return_type_t<RowVec, ColVec>>`. The issue here is that a vector of vars can be allowed here if the other type is complex.

File: test/unit/math/mix/fun/multiply_test.cpp
Comment (SteveBronder) [reply test/unit/math/mix/fun/multiply_test.cpp:238]: I added these

**Comment by bob-carpenter** [comment]
Super cool!  Let me know if you could use another set of eyes on this.  Otherwise, I'm going to get busy writing the complex spec for the language.

**Comment by rok-cesnovar** [comment]
> Otherwise, I'm going to get busy writing the complex spec for the language.

Nice! 

I think other than that we then need a Math issue for all the utility functions like to_complex() for scalar, vectors, matrices and things like `ì()`.

**Comment by bbbales2** [comment]
```
cc1plus.exe: out of memory allocating 65536 bytes
```

is this the error that happens when we need to split tests?

**Comment by rok-cesnovar** [comment]
Yes sir.

**Comment by bbbales2** [comment]
Cool I'll do that real quick.

**Review by bbbales2 — APPROVED** [review]

**Pull request merged**
